File: app/src/main/java/com/mandarin/bcu/androidutil/fakeandroid/BMBuilder.java
package com.mandarin.bcu.androidutil.fakeandroid;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import common.system.fake.FakeImage;
import common.system.fake.ImageBuilder;

public class BMBuilder extends ImageBuilder {
    @Override
    public FIBM build(Object o) {
        if(o == null)
            return null;
        if(o instanceof Bitmap)
            return new FIBM((Bitmap)o);
        if(o instanceof FIBM)
            return (FIBM) o;

        Bitmap b = null;

        if(o instanceof File)
            b = BitmapFactory.decodeFile(((File) o).getAbsolutePath());
        else if (o instanceof byte[]) {
            b = BitmapFactory.decodeByteArray((byte[])o,0,((byte[]) o).length);
        }
        return b == null ? null:new FIBM(b);
    }

    @Override
    public boolean write(FakeImage img, String fmt, Object o) throws IOException {
        Bitmap b = (Bitmap)img.bimg();

        if(b == null)
            return false;
        if(o instanceof File) {
            OutputStream os = new FileOutputStream((File)o);
            return b.compress(Bitmap.CompressFormat.PNG,100,os);
        }
        if(o instanceof OutputStream)
            return b.compress(Bitmap.CompressFormat.PNG,100,(OutputStream) o);
        return false;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/fakeandroid/BMBuilder.kt
package com.mandarin.bcu.androidutil.fakeandroid

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import common.system.fake.FakeImage
import common.system.fake.ImageBuilder
import java.io.*
import java.util.function.Supplier

class BMBuilder : ImageBuilder<Bitmap>() {
    companion object {
        val option = BitmapFactory.Options().apply {
            inScaled = true
            inDensity = inTargetDensity
        }
    }

    @Throws(IOException::class)
    override fun write(img: FakeImage, fmt: String, o: Any): Boolean {
        val b = img.bimg() as Bitmap

        if (o is File) {
            val os: OutputStream = FileOutputStream(o)
            return b.compress(Bitmap.CompressFormat.PNG, 100, os)
        }

        return if (o is OutputStream)
            b.compress(Bitmap.CompressFormat.PNG, 100, o)
        else
            false
    }

    override fun build(f: File?): FakeImage {
        f ?: return FIBM()

        val b = BitmapFactory.decodeFile(f.absolutePath, option)

        return FIBM(b)
    }

    override fun build(sup: Supplier<InputStream>?): FakeImage {
        sup ?: return FIBM()

        val ins = sup.get()

        val b = BitmapFactory.decodeStream(ins, null, option) ?: return FIBM()

        return FIBM(b)
    }

    override fun build(o: Bitmap?): FakeImage {
        o ?: return FIBM()

        return FIBM(o)
    }

    override fun build(w: Int, h: Int): FakeImage {
        return FIBM(Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888))
    }
}